### ISSUE TITLE
Voicemail expiration date is only 12 hours!

### ISSUE THREAD

**Issue by kourosh-samia** [amazon-connect/voicemail-for-amazon-connect#10]
We implemented the Amazon Voicemail last week and noticed that the link we receive in the voicemail email even it says for example it will be expired in 2 days but the token will expire in 12 hours. 
This will defeat the whole purpose of the voicemail system. The audio file exist on S3 how ever the link in the email doesn't take you to the audio file due to expired token.How can this be fixed? Below is the example. 

voicemail email :Voicemail Expiration Date: Sun Apr 19 2020 21:50:15 GMT+0000 (Coordinated Universal Time)

Click Here to listen to the voicemail
You click on the link and get below page a little over 12 hours. Before 12 hours the token is valid and you can here the audio
<Error>
<Code>AccessDenied</Code>
<Message>Request has expired</Message>
<Expires>2020-04-19T21:50:15Z</Expires>
<ServerTime>2020-04-20T18:57:44Z</ServerTime>
<RequestId>0BFD7E0C9300A713</RequestId>
<HostId>
lxNtvqnwABMkbKU8OLvs0llE7BvsFOEU/ikV5orN7njCqBAVWEH9xBNAWbWvjHNpIPlbUUYA0YI=
</HostId>
</Error>

**Comment by efh365** [comment]
Hi, Thank you for reporting this. I have been looking into this and I think I have a fix, but need to wait another 12 hours to verify it. I learned that the problem is the URL expires when the IAM role expires, which is at 12 hours. We should be able to get around this by using an IAM user with SigV4 instead. I will update this once I have a fix and will tell you how to implement it. We will then update the solution with the change.

**Comment by efh365** [comment]
I verified I fix for this issue. Please follow the steps below.

1) In the /src/service/s3.service.js file you need to update the creation of the S3 object to include the access key and secret key of an IAM user and to use sigv4:

```javascript
  constructor() {
        this.s3 = new AWS.S3({
            signatureVersion: 'v4',
            credentials: {
                secretAccessKey: process.env.SECRET_KEY,
                accessKeyId: process.env.ACCESS_KEY
            }
        });
    }
```

2) You will need to create an IAM user that has the same policy as the one that is used by the Lambda function. When you create the IAM user, grant it programmatic access and take the access and secret keys and pass it to the Lambda function as environment variables. If you don't want the secret and access keys to be visible in the environment variables then you can store and access them via AWS Secrets Manager.

3) Build the code and take the generated aws-connect-vm.zip file and upload it to the Lambda function with "ContactVoicemailStream" in the name.

Let me know if you need any more guidance here. We are working on updating the solution to account for this.

**Comment by Donofritz** [comment]
Perhaps there is another way instead of creating an IAM user?   

Instead of using an IAM user with API access in the constructor, another method may be to set the ChainableTemporaryCredentials to a value that corresponds with the Lambda parameter for expiration.  
Note the following below from this reference: 
https://advancedweb.hu/limit-permissions-with-roles-for-signed-urls/ 

So essentially you could leverage the Lambda parameter to calculate the expiration window in the credential to correspond to the appropriate timeframe you want the signed URL to expire.   



Sample code from the referenced URL:

const credentials = new AWS.ChainableTemporaryCredentials({
         params: {RoleArn: roleArn}
});
credentials.expiryWindow = 15 * 60; // 15 minutes

const s3 = new AWS.S3({credentials});

**Comment by InfeCtlll3** [comment]
I believe the limitation is due to s3 preassignedURL token signature limitation as stated here[1] and not a matter of expiryWindow of credentials.

The steps mentioned by efh365@ will guarantee the preassigned url to last up to 7 days.

The steps efh365@ mentioned in short are:
1- At the IAM dashboard, manually create a new programatic user, attach the same policies/permissions as the Lambda role and copy the Secret and Access key.

2- Modify the Lambda code to the one mentioned by the previously mentioned fix, zip it and update the code for the voice mail lambda function.

3- Go to the Lambda console, select the voice mail lambda, under the "Environment variables" edit and create 2 environment variables SECRET_KEY and ACCESS_KEY then set their values to the according secret/access of the newly created user.

ps: Values set in the environment variables will be encrypted with the default KMS key of the account while in transit. In short this means that these values won't be stored in plain text in the backend. However for more security, I'd also advise to use secrets manager as also mentioned by efh365@.

[1] https://aws.amazon.com/premiumsupport/knowledge-center/presigned-url-s3-bucket-expiration/

**Comment by ivassallo19** [comment]
Hey team,

Our company is deploying Voicemail for Amazon Connect, and we've caught this issue as well.
We'd like to implement this code change, but have a few questions.

If the codebase for this changes (by Amazon or the vm team) then our codebase would be out of date, until we redo this action?

And the instructions aren't very clear on how to actually do this. Trying to follow efh365's comment coupled with InfeCtlll3's.
When making the code change in the s3.service.js file, re-zip all files, where do I upload it?
When you say select the voicemail Lambda for the environment variables, which one exactly? I have roughly 16 Lambda functions created automatically.

Thanks in advance.

**Comment by dfw100** [comment]
Hi ivassallo19,

1. Create aws-connect-vm.zip file and IAM user as described by efh365 above.
2. Navigate to https://console.aws.amazon.com/lambda/ 
3. Filter by keyword 'ContactVoicemailStream' to find the Lambda function with description "AWS Lambda Function that will be triggered when records are written into the ContactVoicemail Table."
4. Click on the function name to open the lambda function.
5. Click the 'Upload' button in the Function code section on the configuration page.
6. Select the aws-connect-vm.zip file and upload it to the Lambda function.
7. Add two new environment variables SECRET_KEY and ACCESS_KEY and get their values from the IAM  user created in step 1 above.
8. Save the Lambda function.

Please let us know if you need further assistance.

**Comment by pulverscott** [comment]
Hello All,

I'm trying to do this as well. I'm confused on generating the aws-connect-vm.zip file. Is that a file you get after you run the build commands on the README.md instructions for the project? Or is that a zip of all the .js files in that folder after editing the s3.service.js file?

Some instruction help would be greatly appreciated! Thanks

**Comment by dfw100** [comment]
> Hello All,
> 
> I'm trying to do this as well. I'm confused on generating the aws-connect-vm.zip file. Is that a file you get after you run the build commands on the README.md instructions for the project? Or is that a zip of all the .js files in that folder after editing the s3.service.js file?
> 
> Some instruction help would be greatly appreciated! Thanks

Hi Scott,

aws-connect-vm.zip file is generated at /src/Voicemail-for-amazon-connect/deployment/regional-s3-assets/aws-connect-vm-serverless after running build-s3-dist.sh as described in the README . 

Please let us know if you have further questions.

**Comment by zoogati** [comment]
Hey,

Have been trying to get this built for quite some time. I don't get any generated files for vm-serverless.

**Comment by swalmsley11** [comment]
Hi all,

I've attempted running the build-s3-dist.sh a few times both on my local dev environment and even in a clean cloud9 instance and the build doesn't generate any of the files mentioned. Upon completion, the /src/voicemail-for-amazon-connect/deployment/regional-s3-assets/vm-serverless/ directory is empty, as is the /src/voicemail-for-amazon-connect/source/.dist/aws-connect-vm-serverless/ directory. Below are a few entries I noticed in the console as the build was running.

This occurred right before the "Creating an optimized production build..." message:

`build.sh: line 82: mvn: command not found`
`cp: cannot stat ‘./aws-connect-vm-serverless/target/aws-connect-vm-java.jar’: No such file or directory`
`cp: cannot stat ‘./aws-connect-vm-serverless/.serverless/aws-connect-vm.zip’: No such file or directory`
`cp: cannot stat ‘./aws-connect-vm-serverless/.serverless/cloudformation-template-update-stack.json’: No such file or directory`
`Traceback (most recent call last):`
`  File "./tools/transform.py", line 170, in <module>`
`    transform_template(args.template, args.save, args.zip, args.jar)`
`  File "./tools/transform.py", line 20, in transform_template`
`    with open(template) as json_file:`
`FileNotFoundError: [Errno 2] No such file or directory: '../deployment/aws-connect-vm.yaml'`

And this occurs at the end of the build process:

`cp: omitting directory ‘./.dist/aws-connect-vm-serverless/templates’`
`------------------------------------------------------------------------------`
`[Packing] Templates`
`------------------------------------------------------------------------------`
`cp /home/ec2-user/environment/new/voicemail-for-amazon-connect/deployment/*.template /home/ec2-user/environment/new/voicemail-for-amazon-connect/deployment/global-s3-assets/`
`copy yaml templates and rename`
`cp: cannot stat ‘/home/ec2-user/environment/new/voicemail-for-amazon-connect/deployment/*.yaml’: No such file or directory`
`mv: cannot stat ‘*.yaml’: No such file or directory`
`Updating code source bucket in template with eclipsevoice`

Any assistance would be much appreciated.
Thanks!
